### PR TITLE
Disnake migration

### PR DESCRIPTION
**Pull request by ChrisLovering**
This migrates the bot to disnake, as per recent discussions.

What this includes:
- Changes all references from discord.py to disnake, including in tests
- Bump version of bot-core so discord.py wasn't required as a sub dep
- Use monkey patches from bot-core
- Fix the single breaking change I could find
    - The internal bot.http.edit_message function now has a required `files` kwarg, which is only used in the help channels cog.
    - Set this to None as we never upload files using that message

This not included in this PR:
- Updating the tags to reference disnake. These were left with the discord module, and can be updated later if needed
- Migrating the bot to use all utils that have been moved to bot-core
    - This would inflate this PR even further, so didn't want to pull it into scope.

P.S this is quite a large PR, so reviewing by the commits may be helpful. I tried to keep the huge rename within the single commit https://github.com/python-discord/bot/pull/2103/commits/27b18baaf785091e4e2d5dcdca3ff18981dfcd44, and didn't include anything else in there.

### CODE REVIEW

File: bot/__init__.py
Comment (MarkKoz) [inline bot/__init__.py:19]: This API is kind of silly. If there's only going to be one public function, it would have been better to expose it directly from the utils namespace. You stated yourself that we'd probably never want to apply patches on an individual basis; if patches are no longer needed we'd just remove them entirely from the code base.

File: pyproject.toml
Comment (MarkKoz) [inline pyproject.toml:12]: We basically do versioning already for the change log. Do we plan to use git tags? Then I think this URL could use the tag, which would be clearer than a hash.

**Comment by Bluenix2** [comment]
Just as point of order / reminder, please do not force push in this PR as that will mess with the "Viewed" markers to the right.

**Comment by Bluenix2** [comment]
Good work on this by the way, I ran both Pyright and Mypy in the command line with most type checking disabled (or ignored when reading the output) and I could not find any complaints about `discord` or missing imports (although Mypy failed to find the typings for `disnake.ext.commands` which may make this test incomplete).

**Review by Bluenix2 — CHANGES_REQUESTED** [review]
I found these three references to discord.py still, they should probably be renamed as well:
https://github.com/python-discord/bot/blob/91621bc7ef590118e91bbb2e387db5e14128b530/bot/utils/function.py#L106
https://github.com/python-discord/bot/blob/91621bc7ef590118e91bbb2e387db5e14128b530/bot/utils/function.py#L97
https://github.com/python-discord/bot/blob/91621bc7ef590118e91bbb2e387db5e14128b530/bot/utils/function.py#L139

File: pyproject.toml
Comment (ChrisLovering) [reply pyproject.toml]: Yea, good idea. https://github.com/python-discord/bot/pull/2103/commits/3a03e83776db009494f189183189615d616fdb59

File: bot/__init__.py
Comment (Bluenix2) [reply bot/__init__.py:19]: @MarkKoz I think you should open an issue in the bot-core repository then. I agree with you and would like to see this changed, but at the moment this is not something Chris can change in this PR other than doing `from botcore.utils.monkey_patches import apply_monkey_patches`.

**Comment by Bluenix2** [comment]
Thanks for addressing my comments Chris. I have looked at the diff of each file and not yet found any odd changes - I could also not find any differences between discord.py and disnake when migrating **other than the new text-in-voice feature resulting in some methods returning `VoiceChannel`** which may not include all attributes/methods (`send()` and `edit()` would be present I believe). Any issues that may arise from this that come to mind?

Additionally, this pull request is pretty much incompatible with every other pull request currently open. Has there been any thought put into figuring out which ones should be merged before this is merged, and which ones should be merged afterwards (so that we can let those know)?

File: bot/__init__.py
Comment (ChrisLovering) [reply bot/__init__.py:19]: Raised https://github.com/python-discord/bot-core/pull/35 for this comment, good point Mark

**Comment by ChrisLovering** [comment]
>[ ... ] `VoiceChannel`** which may not include all attributes/methods (`send()` and `edit()` would be present I believe). Any issues that may arise from this that come to mind?
Not that I can think of, we also don't use text in voice channels in the server.

> Additionally, this pull request is pretty much incompatible with every other pull request currently open. Has there been any thought put into figuring out which ones should be merged before this is merged, and which ones should be merged afterwards (so that we can let the author of those know to update to disnake)?

I think any PRs that are open when this is merged will just need to be rebased, I don't think there are any PRs open that should block this.

We of course can extend the offer to help anyone with an open PR to resolve conflicts.

**Comment by ChrisLovering** [comment]
Marked as do not merge to allow me to squash fixup commits on this PR before it hits main

File: bot/__init__.py
Comment (ChrisLovering) [reply bot/__init__.py]: Fixed in https://github.com/python-discord/bot/pull/2103/commits/9e6989aee80ee21a2f3defa94d9eca36834fa771

File: bot/utils/checks.py
Comment (MarkKoz) [inline bot/utils/checks.py:138]: Reviewing every line paid off
```suggestion
        # cooldown logic, taken from disnake's internals
```

**Review by MarkKoz — CHANGES_REQUESTED** [review]
I've looked over all the diffs; looks fine except that one comment. I ran the bot and it started up fine. I did my own search for missed references and couldn't find anything.

File: bot/utils/checks.py
Comment (ChrisLovering) [reply bot/utils/checks.py]: hah, nice spot https://github.com/python-discord/bot/pull/2103/commits/445c4a533eb11e9f9c8d8b438c858f7425775d50

**Review by MarkKoz — APPROVED** [review]

**Review by mbaruh — APPROVED** [review]
Ran most commands, seems to work well. Thanks!

**Pull request merged**
